File: 01. DataStructures/01. Linked List/1. Singly-Linked-List/Problems/Remove Nth Node From End of List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* p = head;
        int count = 0; 
        while(p){
            count++;
            p = p->next;
        }
        
        count = count - n;
        if(count == 0){
            return head->next;
        }
        ListNode *q = head;
        ListNode *temp = head->next;
        while(count>1){
            q = temp;
            temp = temp->next;
            count--;
        }
        q->next = temp->next;
        return head;
    }
};

File: 01. DataStructures/01. Linked List/1. Singly-Linked-List/Problems/Remove Nth Node From End of List/Solution.js

var removeNthFromEnd = function(head, n) {
    
    if(!head) return null;
    // if the node is the only node in the list
    if(!head.next && n == 1) {
        head = null;
        return head;
    }
    
    const linkLen = findLen(head);
    const diff = linkLen - n;
    const firstHeadRef = head;
    

    
    if(diff >= 0) {
        let count = 1;
        
        // if n is equal to the  lenght of the linklist.
        if(diff === 0) {
            head = head.next;
            return head;
        }
        while(count < diff) {
            head = head.next;
            count++;
        }
        if(head.next && head.next.next) {
            head.next = head.next.next;            
        } else{
            head.next = null;            
        }
    } else {
        return null;
    }
    
    return firstHeadRef;
};

var findLen = function(node) {
    let count = 0;
    while(node) {
        node = node.next;
        count++;
    }
    
    return count;
}
